### PR TITLE
Re-fix Issue 12901 in/out contracts on struct constructor must require function body

### PR DESCRIPTION
**Pull request by ibuclaw**
Fix as per #7527, but doesn't revert fix in #3664.

### CODE REVIEW

File: src/dmd/dsymbolsem.d [hunk 1]
```diff
@@ -4780,16 +4803,9 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
...
-            (funcdecl.fensure || funcdecl.frequire) &&
-            !(id && funcdecl.isVirtual()))
-        {
-            auto cd = parent.isClassDeclaration();
-            if (!(cd && cd.isAbstract()))
-                funcdecl.error("in and out contracts on non-virtual/non-abstract functions require function body");
-        }
+    Ldone:
+        if (!funcdecl.fbody && !funcdecl.allowsContractWithoutBody())
+            funcdecl.error("in and out contracts on non-virtual/non-abstract functions require function body");
```
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]:
```
Test failed: 
expected:
----
fail_compilation/fail12901.d(11): Error: constructor fail12901.S.this in and out contracts require function body
----
actual:
----
fail_compilation/fail12901.d(11): Error: constructor fail12901.S.this in and out contracts on non-virtual/non-abstract functions require function body
----
```

How about:

```d
if (parent.isStructDeclaration()))
    funcdecl.error("in and out contracts require function body");
else
   funcdecl.error(""in and out contracts on non-virtual/non-abstract functions require function body");
```

File: src/dmd/dsymbolsem.d [hunk 2]
```diff
@@ -569,6 +569,29 @@ const(char)* getMessage(DeprecatedDeclaration dd)
     return dd.msgstr;
 }
 
+// Returns true if a contact can appear with a function body.
```
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]: s/contact/contract/
s/with/without/

File: src/dmd/dsymbolsem.d [hunk 3]
```diff
@@ -569,6 +569,29 @@ const(char)* getMessage(DeprecatedDeclaration dd)
...
+        InterfaceDeclaration id = parent.isInterfaceDeclaration();
+
+        if (!funcdecl.isAbstract() &&
+            (funcdecl.fensure || funcdecl.frequire) &&
+            !(id && funcdecl.isVirtual()))
+        {
+            auto cd = parent.isClassDeclaration();
+            if (!(cd && cd.isAbstract()))
+                return false;
+        }
```
Comment (wilzbach) [inline src/dmd/dsymbolsem.d]: +1
This is a lot better in terms of readability!

**Review by wilzbach — COMMENTED** [review]
Thanks for the quick follow-up!

**Comment by JinShil** [comment]
Please heed [this comment](https://github.com/dlang/dmd/pull/7527#issuecomment-354373982) from @quickfur 

> Huh, issue 12901 is really more of a problem of allowing arbitrary assignments inside the contract body. The original fix was invalid, since it doesn't stop you from writing `this(int n) in { a = n; } do {}`, which is still contract abuse. I'm not sure what the correct fix should be, perhaps require the contracts be pure or something.

File: src/dmd/dsymbolsem.d (hunk 2 above)
Comment (JinShil) [reply src/dmd/dsymbolsem.d]: The function comment header should use DDoc formatting with `Params` and `Returns`.  See https://dlang.org/spec/ddoc.html#standard_srction

File: src/dmd/dsymbolsem.d [hunk 4]
```diff
@@ -569,6 +569,29 @@ const(char)* getMessage(DeprecatedDeclaration dd)
     return dd.msgstr;
 }
 
+// Returns true if a contact can appear with a function body.
+private bool allowsContractWithoutBody(FuncDeclaration funcdecl)
```
Comment (WalterBright) [inline src/dmd/dsymbolsem.d]: the name of the function appears to contradict the `Returns ...` in the function comment.

File: src/dmd/dsymbolsem.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/dsymbolsem.d]: Actually, I think the readability is impaired by the extensive use of `!`. Not a big deal, but I'd De Morgan the sense around so it matches the comment.

File: src/dmd/dsymbolsem.d [hunk 5]
```diff
@@ -1570,7 +1593,7 @@ extern(C++) final class Semantic3Visitor : Visitor
                 }
 
                 // If declaration has no body, don't set sbody to prevent incorrect codegen.
-                if (funcdecl.fbody || funcdecl.fdensure || funcdecl.fdrequire)
+                if (funcdecl.fbody || funcdecl.allowsContractWithoutBody())
```
Comment (WalterBright) [inline src/dmd/dsymbolsem.d]: The test of `funcdecl.fbody` here is redundant with what `allowsContractWithoutBody` does.

File: src/dmd/dsymbolsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/dsymbolsem.d]: The negation in the error message can be rewritten as a positive:
```
Contracts can  only appear without a body when they are virtual interface functions or abstract.
```
:-)

**Comment by ibuclaw** [comment]
> Please heed this comment from @quickfur

I would open a new issue with a better worded title.  The original fix by Kenji does match the issue description.

File: src/dmd/dsymbolsem.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/dsymbolsem.d]: Not if it's private IMO.

File: src/dmd/dsymbolsem.d (hunk 5 above)
Comment (ibuclaw) [reply src/dmd/dsymbolsem.d]: I've turned it into an assert in the function.

File: src/dmd/dsymbolsem.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/dsymbolsem.d]: I think it's appropriate to use the standard sections but not make it a DDoc comment.

**Comment by quickfur** [comment]
The bug title can be (and arguably *should* be) edited.  The complaint really isn't about whether or not to allow bodies in a ctor with contracts, but about being able to initialize aggregate members from inside a contract, which becomes a bug when compiling without contracts.  So the current bug title is misleading, and trying to "fix" the bug according to the title is a wrong solution.  The real problem here is that contracts should not be able to do things like initialize aggregate members.  The part about empty bodies is a red herring.

**Comment by ibuclaw** [comment]
> The bug title can be (and arguably should be) edited.

The bug was closed for nearly 4 years.  No point bringing it back from the dead.  Raise a new issue with a clear description of what should be fixed.

**Comment by JinShil** [comment]
> The bug was closed for nearly 4 years. No point bringing it back from the dead.

Right.  This was broken accidentally when another PR was merged too early.  Let's put things back together first and fix the underlying problem in another PR.  @quickfur, could you please file a separate issue to document the fundamental problem and your proposed solution, so we don't further sidetrack/delay this PR?

**Comment by ibuclaw** [comment]
Rebased. Hopefully no more surprise rebasing needed.

**Review by JinShil — APPROVED** [review]

**Comment by ibuclaw** [comment]
New issue raised for other problem. https://issues.dlang.org/show_bug.cgi?id=18143

**Pull request merged**